controllers/tenant_compile_pipeline_controller.go

```diff
@@ -0,0 +1,124 @@
+package controllers
+
+import (
+	"context"
+	"fmt"
+	"strings"
+
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/log"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+
+	synv1alpha1 "github.com/projectsyn/lieutenant-operator/api/v1alpha1"
+)
+
+// TenantCompilePipelineReconciler reconciles a Tenant object, specifically the `Spec.CompilePipeline` field, updating the corresponding tenant's git repo accordingly.
+type TenantCompilePipelineReconciler struct {
+	client.Client
+	Scheme *runtime.Scheme
+	ApiUrl string
+}
+
+const (
+	CI_VARIABLE_API_URL                     = "COMMODORE_API_URL"
+	CI_VARIABLE_API_TOKEN                   = "COMMODORE_API_TOKEN"
+	CI_VARIABLE_CLUSTERS                    = "CLUSTERS"
+	CI_VARIABLE_PREFIX_CLUSTER_ACCESS_TOKEN = "ACCESS_TOKEN_"
+	SECRET_KEY_API_TOKEN                    = "token"
+	SECRET_KEY_GITLAB_TOKEN                 = "token"
+)
+
+//+kubebuilder:rbac:groups=syn.tools,resources=tenants,verbs=get;list;watch;update;patch;
+//+kubebuilder:rbac:groups=syn.tools,resources=clusters,verbs=get;list;watch;
+//+kubebuilder:rbac:groups=syn.tools,resources=tenants/status,verbs=get;
+
+func (r *TenantCompilePipelineReconciler) Reconcile(ctx context.Context, request ctrl.Request) (ctrl.Result, error) {
+	reqLogger := log.FromContext(ctx)
+	reqLogger.Info("Reconciling Tenant Compile Pipeline")
+
+	tenant := &synv1alpha1.Tenant{}
+
+	err := r.Client.Get(ctx, request.NamespacedName, tenant)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			return reconcile.Result{}, nil
+		}
+		return reconcile.Result{}, err
+	}
+
+	changed := false
+
+	if !tenant.GetCompilePipelineSpec().Enabled {
+		changed = ensureCiVariablesAbsent(tenant)
+	} else {
+		changed = r.ensureCiVariables(tenant) || changed
+	}
+
+	cluster := &synv1alpha1.Cluster{}
+	pipelineStatus := tenant.GetCompilePipelineStatus()
+	for _, clusterName := range pipelineStatus.Clusters {
+
+		nsName := types.NamespacedName{Name: clusterName, Namespace: tenant.GetNamespace()}
+		err := r.Client.Get(ctx, nsName, cluster)
+		if err != nil && !errors.IsNotFound(err) {
+			return reconcile.Result{}, fmt.Errorf("while reconciling CI variables for clusters: %w", err)
+		}
+
+		if err == nil {
```

Comment:
We might want to at least log an error. (Or even better: collect errors using multierr).

Reply:
I think this was resolved by accident?

Reply:
No, I added error logging. I think GitHub is doing some weird caching here o.O

Reply:
Are you sure it's pushed? I've never seen github caching this kinda stuff.

Reply:
Ahh shit sorry i was confused by the structure.

Reply:
See https://github.com/projectsyn/lieutenant-operator/pull/292#discussion_r1691520259